src/debug.cc

```diff
@@ -70,19 +99,315 @@ class DebugFile
     FILE *file_ = nullptr; ///< opened "real" file or nil; never stderr
 };
 
-/// configured cache.log file or stderr
+/// meta-information of a Finish()ed debugs() message
+class DebugMessageHeader
+{
+public:
+    DebugMessageHeader(const DebugRecordCount aRecordNumber, const Debug::Context &context):
+        recordNumber(aRecordNumber),
+        timestamp(getCurrentTime()),
+        section(context.section),
+        level(context.level),
+        forceAlert(context.forceAlert),
+        role_XXX(myLeadingRole_XXX())
+    {
+    }
+
+    DebugRecordCount recordNumber; ///< LogMessage() calls before this message
+    time_t timestamp; ///< approximate debugs() call time
+    int section; ///< debugs() section
+    int level; ///< debugs() level
+    bool forceAlert; ///< debugs() forceAlert flag
+
+    const char *role_XXX; ///< debugs() caller process role
+};
+
+/// a fully processed debugs(), ready to be logged
+class DebugMessage
+{
+public:
+    using Header = DebugMessageHeader;
+    DebugMessage(const Header &aHeader, const std::string &aBody);
+
+    Header header; ///< debugs() meta-information; reflected in log line prefix
+    std::string body; ///< the log line after the prefix (without the newline)
+};
+
+/// debugs() messages captured in LogMessage() call order
+using DebugMessages = std::deque<DebugMessage>;
+
+/// a receiver of debugs() messages (e.g., stderr or cache.log)
+class DebugChannel
+{
+public:
+    using EarlyMessages = std::unique_ptr<DebugMessages>;
+
+    explicit DebugChannel(const char * const aName): name(aName), earlyMessages(new DebugMessages()) {}
+    virtual ~DebugChannel() {}
+
+    // no copying or moving or any kind (for simplicity sake and to prevent accidental copies)
+    DebugChannel(DebugChannel &&) = delete;
+
+    /// whether we are still expecting (and buffering) early messages
+    bool collectingEarlyMessages() const { return bool(earlyMessages); }
+
+    /// end early message buffering, logging any saved messages
+    void stopEarlyMessageCollection();
+
+    /// end early message buffering, without logging any saved messages
+    /// \returns (a possibly empty container with) saved messages or nil
+    EarlyMessages releaseEarlyMessages() { return EarlyMessages(earlyMessages.release()); }
+
+    /// Log the message to the channel if the channel accepts (such) messages.
+    /// This logging may be delayed until the channel configuration is settled.
+    virtual void log(const DebugMessageHeader &, const std::string &body) = 0;
+
+protected:
+    /// stores the given early message (if possible) or forgets it (otherwise)
+    /// \returns whether the message was stored
+    bool saveMessage(const DebugMessageHeader &header, const std::string &body);
+
+    /// log() all (saved but not yet written "early") messages from this channel
+    /// and their channel (if given), in recordNumber order
+    void logSavedAnd(DebugChannel *theirChannelOrNil = nullptr);
```

Comment:
> did Squid log those early message to syslog

yes.

Reply:
> did Squid log those early message to syslog

The Squid before d7ca82e logged early DBG_CRITICAL messages to Syslog immediately, including messages written before openlog() (syslog() [works](https://www.gnu.org/software/libc/manual/html_node/openlog.html) even without openlog()). With the PR changes, we buffer early messages before openlog(). This is the difference: in the case of SIGSEGV the buffered CRITICAL messages are lost (since the handler is assigned later in MainInitialize()), while the old Squid successfully writes them to Syslog (before it dies).

> if yes, fixing that is in scope

In order to reconcile the old behavior with the new functionality, I think we should adjust the PR code to buffer IMPORTANT messages only (and syslog() the CRITICAL ones).

Reply:
To allow writing of critical and ForceAlert early messages without dropping all earlier messages, SyslogChannel::log() may need to do something similar to the sketch below:

```C++
if (header.recordNumber <= lastWrittenRecordNumber)
    return;

if (!shouldWrite()) {
    if (earlyMessages)
        (void)saveMessage(header, body);
    return;
}

// Do not stopEarlyMessageCollection() here: The already saved earlier 
// messages are doomed, but future early messages still have a chance.

syslog();
noteWritten();
```

StderrChannel::log() has to make a similar decision but it calls stopEarlyMessageCollection() when logging the very first message instead. I am not sure stderr code is wrong, but it would be nice to explain the difference between the two log() methods if that difference is meaningful.

Reply:
I followed your sketch at 08eccce (syslog) and 6014f02 (stderr).
With these changes, both syslog and stderr dump CRITICAL messages immediately, but

* Syslog continues collecting early messages (if still needed).
* Stderr stops early messages collecting, dumping them.

IIRC we discussed that a CRITICAL message may be treated as a kind of 'emergency' situation, which should cause stop collecting early messages (and dump the available ones). Should not we make both syslog and stderr behave in the same way? On one hand, with this approach (as in stderr) IMPORTANT early messages are lost after the first CRITICAL one. On the other hand, if we continue collecting early messages (as in syslog), we lose all messages preceding the first CRITICAL one. So, having different approaches for stderr and syslog, we can (theoretically) manually combine messages from both sources (thus avoiding message loss). With identical approaches, we cannot manually restore the message layout. Note that this discussion is probably a bit 'far fetched' because there are none (or maybe a few) early CRITICAL messages in Squid now.

Also I refactored DebugChannel::saveMessage(), removing the unused return value.

Reply:
The issue of logging early critical messages may sound messy, but I think we can make most decisions here based on the existing guiding principles:

* stderr: Logging early would violate many Squid configurations (because stderr logging is not enabled by default), so we should _not_ log critical messages to stderr early (just because they are critical). In case of an extremely rare early segfault, the saved messages will not be logged, but fixing such a segault most likely requires core dump analysis, and that analysis can recover any saved messages. It does not make sense to violate many configurations just to improve support for a very rare event (early segfault) that arguably has enough support already.

* syslog: Based on your research, Squid was syslogging all critical messages before our PRs. We have no good reason to change that part of Squid interface AFAICT -- it is easily supported by the new code design. We should log early critical syslog messages.
If we have to log a critical syslog message, should we also log important messages that precede them, instead of effectively hiding them from syslog (by logging a subsequent critical message)? I think the answer to that question is "no, we should not log important early syslog messages preceding critical messages" for the same reasons as those discussed in the stderr bullet -- logging of those important messages to syslog would violate many Squid configurations, and early messages leading to a segfault can be recovered via core dump analysis.

Do you agree with the above logic? If yes, does not current PR code match the above answers/decisions?

Squid should schedule custom segfault handlers as early as possible, and probably earlier than they are scheduled now, but improving that scheduling is not trivial (because the handling depends on command line options) and, hence, that topic/improvements can and should be viewed as being outside this PR scope IMO.

Reply:
> Alex: If yes, does not current PR code match the above answers/decisions?

I think your summary implied that (some of the) stderr changes should be undone, but I have not reviewed the code yet. If they should be undone and/or if other simple changes are needed, please implement them (assuming you agree with my conclusions).

Reply:
I agree that stderr should not show any messages until/unless it is configured to do that. (at least the admin would excect this behavior). The other point is that the master Squid printed CRITICAL messages immediately. This is a minor problem since there are few (perhaps none) such messages. So let's do as you suggested. 

>  does not current PR code match the above answers/decisions

It should match these decisions after b06bddf adjustment.